cdap-app-fabric/src/test/java/co/cask/cdap/internal/profile/ProfileMetricScheduledServiceTest.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.profile;
+
+import co.cask.cdap.api.dataset.lib.cube.AggregationFunction;
+import co.cask.cdap.api.dataset.lib.cube.TimeValue;
+import co.cask.cdap.api.metrics.MetricDataQuery;
+import co.cask.cdap.api.metrics.MetricStore;
+import co.cask.cdap.api.metrics.MetricTimeSeries;
+import co.cask.cdap.api.metrics.MetricsCollectionService;
+import co.cask.cdap.common.app.RunIds;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.utils.Tasks;
+import co.cask.cdap.internal.AppFabricTestHelper;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.id.ProfileId;
+import co.cask.cdap.proto.id.ProgramRunId;
+import com.google.common.collect.ImmutableMap;
+import com.google.inject.Injector;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+public class ProfileMetricScheduledServiceTest {
+  private static Injector injector;
+
+  @BeforeClass
+  public static void setupClass() {
+    injector = AppFabricTestHelper.getInjector();
+  }
+
+  @Test
+  public void testProfileMetrics() throws Exception {
+    ProgramRunId runId = NamespaceId.DEFAULT.app("myApp").workflow("myProgram").run(RunIds.generate());
+    ProfileId profileId = NamespaceId.DEFAULT.profile("myProfile");
+    MetricsCollectionService collectionService = injector.getInstance(MetricsCollectionService.class);
+    MetricStore metricStore = injector.getInstance(MetricStore.class);
+
+    // There are 60 nodes, we emit the metrics each 2 seconds, so each time the node minute should go up by 2 min
+    ProfileMetricScheduledService scheduledService = new ProfileMetricScheduledService(collectionService, runId,
+                                                                                       profileId, 60, 2);
+    scheduledService.startAndWait();
+    // sleep for 6 seconds and expect 3 increment about the metrics
+    TimeUnit.SECONDS.sleep(6);
```

Comment:
most of the time a sleep in a test means something can be improved. I think you can just add a package private method to the service that does the actual metric writing, and call it in runOneIteration() and in this test. That way you test your actual logic, and not the fact that guava's ScheduledService works.

Reply:
Fixed